File: app/src/FizzBuzzType.ts
export abstract class FizzBuzzType {
  abstract generate(number: number): string

  protected fizz(number: number): boolean {
    return number % 3 === 0
  }

  protected buzz(number: number): boolean {
    return number % 5 === 0
  }
}

export class FizzBuzzType01 extends FizzBuzzType {
  generate(number: number): string {
    if (this.fizz(number) && this.buzz(number)) {
      return 'FizzBuzz'
    } else if (this.fizz(number)) {
      return 'Fizz'
    } else if (this.buzz(number)) {
      return 'Buzz'
    }

    return number.toString()
  }
}

export class FizzBuzzType02 extends FizzBuzzType {
  generate(number: number): string {
    return number.toString()
  }
}

export class FizzBuzzType03 extends FizzBuzzType {
  generate(number: number): string {
    if (this.fizz(number) && this.buzz(number)) {
      return 'FizzBuzz'
    } else {
      return number.toString()
    }
  }
}

File: app/fizz_buzz_type.cpp
#include "fizz_buzz_type.h"

std::string FizzBuzzType01::generate(int number) {
    if (number % 3 == 0 && number % 5 == 0) {
        return "FizzBuzz";
    }
    if (number % 3 == 0) {
        return "Fizz";
    }
    if (number % 5 == 0) {
        return "Buzz";
    }
    return std::to_string(number);
}

std::string FizzBuzzType02::generate(int number) {
    return std::to_string(number);
}

std::string FizzBuzzType03::generate(int number) {
    if (number % 3 == 0 && number % 5 == 0) {
        return "FizzBuzz";
    }
    return std::to_string(number);
}